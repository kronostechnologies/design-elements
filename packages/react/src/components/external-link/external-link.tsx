import React, { MouseEvent, ReactElement, useCallback } from 'react';

import { Icon, IconName } from '../icon/icon';
import { StyledLink } from '../route-link/styles/styled-link';

interface ExternalLinkProps {
    className?: string;
    disabled?: boolean;
    href?: string;
    iconName?: IconName;
    label?: string;
    target?: string;

    onClick?(): void;
}

export function ExternalLink({
    className, disabled, href = '', iconName, label, onClick, target,
}: ExternalLinkProps): ReactElement {
    const handleClick: (event: MouseEvent<HTMLAnchorElement>) => void = useCallback((event) => {
        if (!href) {
            event.preventDefault();
        }
        onClick?.();
    }, [href, onClick]);

    return (
        <StyledLink
            aria-disabled={disabled ? 'true' : 'false'}
            className={['external', className, !label && 'iconOnly'].filter(Boolean).join(' ')}
            disabled={disabled}
            href={disabled ? undefined : href}
            onClick={disabled ? undefined : handleClick}
            target={target}
        >
            {iconName && <Icon name={iconName} size="16" />}
            {label}
        </StyledLink>
    );
}
