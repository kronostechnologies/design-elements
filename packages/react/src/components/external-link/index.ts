export { ExternalLink } from './external-link';
export type { ExternalLinkProps } from './external-link';
