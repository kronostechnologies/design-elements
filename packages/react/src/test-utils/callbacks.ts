export function doNothing(): void {
    // do nothing
}
