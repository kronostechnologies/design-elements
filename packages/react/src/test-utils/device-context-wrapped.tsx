import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
