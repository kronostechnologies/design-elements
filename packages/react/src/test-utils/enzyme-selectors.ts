import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
