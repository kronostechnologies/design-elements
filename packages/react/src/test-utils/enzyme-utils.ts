import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
