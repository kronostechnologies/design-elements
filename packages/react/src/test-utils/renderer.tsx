import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
