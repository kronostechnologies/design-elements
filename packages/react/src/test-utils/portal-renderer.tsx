import { render, RenderResult } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { DeviceType } from '../components/device-context-provider/device-context-provider';
import { AllProviders } from './renderer';

export function renderPortalWithProviders(
    component: ReactElement,
    device?: DeviceType,
): RenderResult {
    return render(<AllProviders device={device}>{component}</AllProviders>);
}
