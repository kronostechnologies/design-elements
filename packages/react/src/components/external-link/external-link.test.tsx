import { mountWithProviders, renderWithProviders } from '@design-elements/test-utils/renderer';
import React from 'react';
import { ExternalLink } from './external-link';

describe('External Link', () => {
    test('onClick callback is called when clicked', () => {
        const callback = jest.fn();
        const wrapper = mountWithProviders(
            <ExternalLink onClick={callback} href="#" label="External Link" />,
        );

        wrapper.find(ExternalLink).simulate('click');
        expect(callback).toHaveBeenCalledTimes(1);
    });
    test('matches snapshot', () => {
        const tree = renderWithProviders(
            <ExternalLink href="https://www.google.ca/" label="External Link" />,
        );

        expect(tree).toMatchSnapshot();
    });
    test('with icon matches snapshot', () => {
        const tree = renderWithProviders(
            <ExternalLink href="#" label="External Link" iconName="mail" />,
        );

        expect(tree).toMatchSnapshot();
    });
    test('without href matches snapshot', () => {
        const tree = renderWithProviders(
            <ExternalLink label="External Link" iconName="mail" />,
        );

        expect(tree).toMatchSnapshot();
    });
    test('disabled matches snapshot', () => {
        const tree = renderWithProviders(
            <ExternalLink href="#" label="External Link" disabled />,
        );

        expect(tree).toMatchSnapshot();
    });
});
