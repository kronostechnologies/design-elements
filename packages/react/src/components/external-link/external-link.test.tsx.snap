// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`External Link matches snapshot (disabled) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #84C6EA;
  cursor: default;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.875rem;
  line-height: 1.5rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:focus {
  outline: none;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c0:focus:not(:focus-visible) {
  box-shadow: none;
}

.c0:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c4 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c3 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 1rem;
  margin-left: var(--spacing-half);
  margin-right: 0;
  width: 1rem;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1:visited svg {
  color: #62a;
}

.c2 {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

<a
  aria-disabled="true"
  class="c0 c1"
  disabled=""
  target="_blank"
>
  <span
    class="c2"
  >
    External Link
  </span>
  <svg
    aria-label="opens in a new tab"
    class="c3"
    color="currentColor"
    focusable="false"
    height="16"
    role="img"
    width="16"
  />
  <span
    class="c4"
    data-testid="screen-reader-text"
  >
    (opens in a new tab)
  </span>
</a>
`;

exports[`External Link matches snapshot (label and icon) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #006296;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.875rem;
  line-height: 1.5rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:hover {
  color: #003A5A;
}

.c0:visited {
  color: #62a;
}

.c0:focus {
  outline: none;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c0:focus:not(:focus-visible) {
  box-shadow: none;
}

.c0:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c5 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin-right: var(--spacing-quarter);
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 1rem;
  margin-left: var(--spacing-half);
  margin-right: 0;
  width: 1rem;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1:visited svg {
  color: #62a;
}

.c3 {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

<a
  aria-disabled="false"
  class="c0 c1"
  href="#"
  target="_blank"
>
  <svg
    aria-hidden="true"
    class="c2"
    color="currentColor"
    focusable="false"
    height="16"
    width="16"
  />
  <span
    class="c3"
  >
    External Link
  </span>
  <svg
    aria-label="opens in a new tab"
    class="c4"
    color="currentColor"
    focusable="false"
    height="16"
    role="img"
    width="16"
  />
  <span
    class="c5"
    data-testid="screen-reader-text"
  >
    (opens in a new tab)
  </span>
</a>
`;

exports[`External Link matches snapshot (only icon) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #006296;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.875rem;
  line-height: 1.5rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:hover {
  color: #003A5A;
}

.c0:visited {
  color: #62a;
}

.c0:focus {
  outline: none;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c0:focus:not(:focus-visible) {
  box-shadow: none;
}

.c0:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c5 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin-right: var(--spacing-quarter);
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 1rem;
  margin-left: var(--spacing-half);
  margin-right: 0;
  width: 1rem;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1:visited svg {
  color: #62a;
}

.c3 {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

<a
  aria-disabled="false"
  class="c0 c1"
  href="#"
  target="_blank"
>
  <svg
    aria-hidden="true"
    class="c2"
    color="currentColor"
    focusable="false"
    height="16"
    width="16"
  />
  <span
    class="c3"
  />
  <svg
    aria-label="opens in a new tab"
    class="c4"
    color="currentColor"
    focusable="false"
    height="16"
    role="img"
    width="16"
  />
  <span
    class="c5"
    data-testid="screen-reader-text"
  >
    (opens in a new tab)
  </span>
</a>
`;

exports[`External Link matches snapshot (without href) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #006296;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.875rem;
  line-height: 1.5rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:hover {
  color: #003A5A;
}

.c0:visited {
  color: #62a;
}

.c0:focus {
  outline: none;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c0:focus:not(:focus-visible) {
  box-shadow: none;
}

.c0:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c5 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin-right: var(--spacing-quarter);
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 1rem;
  margin-left: var(--spacing-half);
  margin-right: 0;
  width: 1rem;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1:visited svg {
  color: #62a;
}

.c3 {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

<a
  aria-disabled="false"
  class="c0 c1"
  href=""
  target="_blank"
>
  <svg
    aria-hidden="true"
    class="c2"
    color="currentColor"
    focusable="false"
    height="16"
    width="16"
  />
  <span
    class="c3"
  >
    External Link
  </span>
  <svg
    aria-label="opens in a new tab"
    class="c4"
    color="currentColor"
    focusable="false"
    height="16"
    role="img"
    width="16"
  />
  <span
    class="c5"
    data-testid="screen-reader-text"
  >
    (opens in a new tab)
  </span>
</a>
`;

exports[`External Link matches snapshot 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #006296;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.875rem;
  line-height: 1.5rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:hover {
  color: #003A5A;
}

.c0:visited {
  color: #62a;
}

.c0:focus {
  outline: none;
}

.c0:focus {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c0:focus:not(:focus-visible) {
  box-shadow: none;
}

.c0:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #84C6EA;
  box-shadow: 0 0 0 1px #006296,0 0 0 3px #84C6EA;
}

.c4 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c3 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 1rem;
  margin-left: var(--spacing-half);
  margin-right: 0;
  width: 1rem;
}

.c1 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1:visited svg {
  color: #62a;
}

.c2 {
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
}

<a
  aria-disabled="false"
  class="c0 c1"
  href="https://www.google.ca/"
  target="_blank"
>
  <span
    class="c2"
  >
    External Link
  </span>
  <svg
    aria-label="opens in a new tab"
    class="c3"
    color="currentColor"
    focusable="false"
    height="16"
    role="img"
    width="16"
  />
  <span
    class="c4"
    data-testid="screen-reader-text"
  >
    (opens in a new tab)
  </span>
</a>
`;
